segmentation/include/pcl/segmentation/extract_clusters.h

```diff
@@ -41,10 +41,148 @@
+  * if that is not the case then things can go very wrong. For speed reasons, we only check the sizes and not the content
+  * \ingroup segmentation
+  */
+  template <typename PointT, typename FunctorT> void
+   extractEuclideanClusters(
+      ConstCloudIterator<PointT> &it, const PointCloud<PointT> &cloud, FunctorT additional_filter_criteria,
+      const typename search::Search<PointT>::Ptr &tree, float tolerance, std::vector<PointIndices> &clusters,
+      unsigned int min_pts_per_cluster = 1, unsigned int max_pts_per_cluster = std::numeric_limits<int>::max())
+  {
+    if (tree->getInputCloud ()->points.size () != cloud.points.size ())
+    {
+      PCL_ERROR ("[pcl::extractEuclideanClusters] Tree built with a different point cloud size (%lu) than the input cloud (%lu)!\n", tree->getInputCloud ()->points.size (), cloud.points.size ());
+      return;
+    }
+
+    // Check if the tree is sorted -- if it is we don't need to check the first element
+    index_t nn_start_idx = tree->getSortedResults () ? 1 : 0;
+    // Create a bool vector of processed point indices, and initialize it to false
+    std::vector<bool> processed (cloud.points.size (), false);
+
+    for (; it.isValid(); ++it) {
+      if (processed[it.getCurrentIndex()])
+        continue;
+
+      index_t sq_idx = 0;
+      clusters.emplace_back();
+      auto& seed_queue = clusters.back();
+      seed_queue.indices.push_back (it.getCurrentIndex());
+
+      processed[it.getCurrentIndex()] = true;
+
+      while (sq_idx < static_cast<index_t> (seed_queue.indices.size()))
+      {
+        Indices nn_indices;
+        std::vector<float> nn_distances;
+
+        // Search for sq_idx
+        if (!tree->radiusSearch (seed_queue.indices[sq_idx], tolerance, nn_indices, nn_distances))
+        {
+          sq_idx++;
+          continue;
+        }
+
+        for (index_t j = nn_start_idx; j < nn_indices.size (); ++j) // can't assume sorted (default isn't!)
+        {
+          if (processed[nn_indices[j]]) // Has this point been processed before ?
+            continue;
+
+          if (additional_filter_criteria(it.getCurrentIndex(), j, nn_indices)) {
+            seed_queue.indices.push_back(nn_indices[j]);
+            processed[nn_indices[j]] = true;
+          }
+        }
+
+        sq_idx++;
+      }
+
+      // If this queue is satisfactory, add to the clusters
+      if (seed_queue.indices.size () >= min_pts_per_cluster && seed_queue.indices.size () <= max_pts_per_cluster)
+        seed_queue.header = cloud.header;
+      else
+        clusters.pop_back();
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+  /** \brief Decompose a region of space into clusters based on the Euclidean distance between points
+    * \param cloud the point cloud message
+    * \param additional_filter_criteria functor which is used as an additonal criteria that all points being added to the cluster must satisfy
```

Comment:
What's the constraints on the functor? Does `[]{return true;}` suffice? Will `[](MyClass ...A){return AnotherClass{A...};}` work?

Tighten the interface appropriately:
* enough leeway for user
* no need to read the documentation based on error message

Reply:
Done. Pick it up form https://github.com/PointCloudLibrary/pcl/blob/c262ec63b1e95bf629b1011a25e87df0ade4e9f1/filters/include/pcl/filters/functor_filter.h#L16-L21
Slighlty off topic, whats the reason for using `pcl::remove_cvref_t` and then adding `const` and `&`.

Reply:
That's a brain-fart. The remove_cvref was meant for something else, but in refactoring from old version to the newer simplified checker implementation, it managed to slip through.